core/storage/page_bitmap.rs

```diff
@@ -0,0 +1,279 @@
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    /// Convert page index to word index and bit offset
+    const fn page_to_word_and_bit(page_idx: u32) -> (usize, u32) {
+        (
+            (page_idx >> Self::WORD_SHIFT) as usize,
+            page_idx & Self::WORD_MASK,
+        )
+    }
+
+    /// Allocates a single free page from the bitmap.
+    ///
+    /// This method scans from high to low addresses to preserve contiguous
+    /// runs of free pages at the low end of the bitmap.
+    pub fn alloc_one(&mut self) -> Option<u32> {
+        if self.n_pages == 0 {
+            return None;
+        }
+
+        // Scan from high to low to preserve contiguous runs at the front
+        for (word_idx, word) in self.words.iter_mut().enumerate().rev() {
+            if *word != Self::ALL_ALLOCATED {
+                // Find the highest set bit (rightmost free page in this word)
+                let bit = 63 - word.leading_zeros();
+                let page_idx = Self::word_and_bit_to_page(word_idx, bit);
+
+                if page_idx < self.n_pages {
+                    *word &= !(1u64 << bit);
+                    // store hint for next single allocation
+                    self.scan_one_high = page_idx.saturating_sub(1);
+                    return Some(page_idx);
+                }
+            }
+        }
+        None
+    }
+
+    /// Allocates a contiguous run of `need` pages from the bitmap.
+    /// This method scans from low to high addresses, starting from the
+    /// `scan_run_low` pointer.
+    pub fn alloc_run(&mut self, need: u32) -> Option<u32> {
+        if need == 1 {
+            return self.alloc_one();
+        }
+        if need == 0 || need > self.n_pages {
+            return None;
+        }
+
+        // Two-pass search with scan_hint optimization
+        let mut search_start = self.scan_run_low;
+        for pass in 0..2 {
+            if pass == 1 {
+                search_start = 0;
+            }
+
+            if let Some(found) = self.find_free_run(search_start, need) {
+                self.mark_run(found, need, false);
+                self.scan_run_low = found + need;
+                return Some(found);
+            }
+
+            if search_start == 0 {
+                // Already searched from beginning
+                break;
+            }
+        }
+        None
+    }
+
+    /// Search for a free run of `need` pages beginning from `start`
+    fn find_free_run(&self, start: u32, need: u32) -> Option<u32> {
+        let mut pos = start;
+        let limit = self.n_pages.saturating_sub(need - 1);
+
+        while pos < limit {
+            if let Some(next_free) = self.next_free_bit_from(pos) {
+                if next_free + need > self.n_pages {
+                    break;
+                }
+                if self.check_run_free(next_free, need) {
+                    return Some(next_free);
+                }
+                pos = next_free + 1;
```

Comment:
Isn't it inefficient to just increment `pos` by 1 if a run of `need` wasn't available for current `pos`?

Pseudo:
```rust
// initially, let's say:
pos: 630, need: 20
// and let's say:
next_free = 666
check_run_free(666, 20) = false
pos = 667 // we know 666 was free, and we increment by 1
next_free = next_free_bit_from(667)
// it might be the case that 667 is free, 668 is free ... 700 is free,
// but still not enough to form a contiguous run of 20
```

Should `check_run_free` return the first page that wasn't free in case the run cannot be fulfilled from `next_free` onwards? and then we could use that for the next iteration of `next_free_bit_from`?